### PR TITLE
Support references

### PR DESCRIPTION
**Pull request by michaelvoronov**
Linked PRs:
https://github.com/fluencelabs/fce/pull/70
https://github.com/fluencelabs/interface-types/pull/8

### CODE REVIEW

File: crates/wit/src/parsed_type.rs [hunk 1]
```diff
@@ -150,74 +124,80 @@ impl ParsedType {
...
+fn type_to_path_passing_style(input_type: &syn::Type) -> syn::Result<(&syn::Path, PassingStyle)> {
+    match input_type {
+        syn::Type::Path(path) => Ok((&path.path, PassingStyle::ByValue)),
+        syn::Type::Reference(type_reference) => match &*type_reference.elem {
+            syn::Type::Path(path) => {
+                // TODO: support lifetimes
+                if type_reference.lifetime.is_some() {
+                    return Err(Error::new(
+                        input_type.span(),
+                        "Lifetime is unsupported at the moment",
```
Comment (folex) [inline crates/wit/src/parsed_type.rs:155]:
```suggestion
                        "Lifetimes are unsupported at the moment",
```

File: crates/wit/src/parsed_type.rs [hunk 2]
```diff
@@ -150,74 +124,80 @@ impl ParsedType {
...
-
-        Ok(())
+// parse generic param T in Vec<T> to syn::Type
+fn parse_vec_bracket(args: &syn::PathArguments) -> syn::Result<&syn::Type> {
+    // checks that T is angle bracketed
+    let generic_arg = match args {
+        syn::PathArguments::AngleBracketed(args) => Ok(args),
+        _ => Err(Error::new(
+            args.span(),
+            "It has to be a bracketed value after Vec",
```
Comment (folex) [inline crates/wit/src/parsed_type.rs:184]:
```suggestion
            "expected value in angle brackets (<>)",
```

File: crates/wit/src/parsed_type.rs [hunk 3]
```diff
@@ -150,74 +124,80 @@ impl ParsedType {
...
+        )),
+    }?;
+
+    let arg = generic_arg.args.first().ok_or_else(|| {
+        Error::new(
+            generic_arg.span(),
+            "Unsuitable type in Vec brackets, lifetimes, bindings, constraints and consts are unsupported",
```
Comment (folex) [inline crates/wit/src/parsed_type.rs:191]:
```suggestion
            "Invalid type in Vec brackets. (NOTE: lifetimes, bindings, constraints and consts are not supported)",
```

File: crates/wit/src/parsed_type.rs [hunk 4]
```diff
@@ -150,74 +124,80 @@ impl ParsedType {
...
+        )
+    })?;
+
+    // converts T to syn::Type
+    match arg {
+        syn::GenericArgument::Type(ty) => Ok(ty),
+        _ => Err(Error::new(
+            arg.span(),
+            "Unsuitable type in Vec brackets, lifetimes, bindings, constraints and consts are unsupported",
```
Comment (folex) [inline crates/wit/src/parsed_type.rs:200]:
```suggestion
            "Invalid type in Vec brackets. (NOTE: lifetimes, bindings, constraints and consts are not supported)",
```

File: crates/wit/src/parsed_type/foreign_mod_prolog.rs [hunk 5]
```diff
@@ -76,11 +77,11 @@ impl ForeignModPrologGlueCodeGenerator for Vec<(String, ParsedType)> {
                 arg_names.push(arg_ident.clone());
 
                 match ty {
-                    ParsedType::Utf8String => {
+                    ParsedType::Utf8String(PassingStyle::ByValue) => {
```
Comment (folex) [inline crates/wit/src/parsed_type/foreign_mod_prolog.rs:80]: `&str` isn't supported?

Comment (folex) [inline crates/wit/src/parsed_type/foreign_mod_prolog.rs:80]: maybe add a comment on the meaning of the matched variants? and on `_ => {}` would be nice too!

File: crates/wit/src/token_stream_generator/record_generator/record_deserializer.rs [hunk 6]
```diff
@@ -103,7 +103,7 @@ impl RecordDeserializerGlueCodeGenerator for fce_ast_types::AstRecordItem {
                         let #field = unsafe { String::from_raw_parts(raw_record[#ptr_id] as _, raw_record[#size_id] as _, raw_record[#size_id] as _) };
                     }
                 }
-                ParsedType::Vector(ty) => {
+                ParsedType::Vector(ty, _) => {
```
Comment (folex) [inline crates/wit/src/token_stream_generator/record_generator/record_deserializer.rs:106]: could you please add a comment on the meaning of the following?
```

                    let ptr_id = value_id;
                    let size_id = value_id + 1;
                    value_id += 1;
```
I'm kinda confused on why `value_id` is incremented here, but not for primitive types like `F32`.

File: crates/wit/src/parsed_type/vector_utils.rs [hunk 7]
```diff
@@ -137,12 +137,12 @@ pub(crate) fn generate_vector_deserializer(
     let arg = crate::new_ident!(arg_name);
 
     let values_deserializer = match value_ty {
-        ParsedType::Boolean => {
+        ParsedType::Boolean(_) => {
             quote! {
                 unimplemented!()
```
Comment (folex) [inline crates/wit/src/parsed_type/vector_utils.rs:142]:
```suggestion
                unimplemented!("Vector of booleans is not supported")
```

File: crates/wit/src/parsed_type/foreign_mod_prolog.rs (hunk 5 above)
Comment (michaelvoronov) [reply crates/wit/src/parsed_type/foreign_mod_prolog.rs:80]: Supported, but they don't need to be explicitly deleted.

File: crates/wit/src/parsed_type/fn_prolog.rs [hunk 8]
```diff
@@ -122,41 +138,31 @@ fn generate_type_lifting_prolog(
...
-        I64(passing_style) => passing_style,
-        F32(passing_style) => passing_style,
-        F64(passing_style) => passing_style,
-        Utf8Str(passing_style) => passing_style,
-        Utf8String(passing_style) => passing_style,
-        Vector(_, passing_style) => passing_style,
-        Record(_, passing_style) => passing_style,
+fn converted_arg_modifier(passing_style: &PassingStyle) -> proc_macro2::TokenStream {
+    match passing_style {
+        PassingStyle::ByMutRef => quote! { mut },
```
Comment (folex) [inline crates/wit/src/parsed_type/fn_prolog.rs:165]: not `quote! { &mut }`?

**Review by folex — APPROVED** [review]
I'm not sure of mem::forget soundness

File: crates/wit/src/parsed_type/fn_prolog.rs (hunk 8 above)
Comment (michaelvoronov) [reply crates/wit/src/parsed_type/fn_prolog.rs:165]: It's used only for `let`

File: crates/wit/src/parsed_type/fn_epilog.rs [hunk 9]
```diff
@@ -24,6 +27,14 @@ pub(crate) struct FnEpilogDescriptor {
     pub(crate) fn_return_type: proc_macro2::TokenStream,
     pub(crate) return_expression: proc_macro2::TokenStream,
     pub(crate) epilog: proc_macro2::TokenStream,
+    pub(crate) mem_forget: proc_macro2::TokenStream,
+}
+
+/// Contains all ingredients needed for epilog creation.
+pub(crate) struct FnEpilogIngredients<'i> {
```
Comment (folex) [inline crates/wit/src/parsed_type/fn_epilog.rs:34]: nice name :)

File: crates/wit/src/parsed_type/vector_utils.rs [hunk 10]
```diff
@@ -180,8 +177,11 @@ pub(crate) fn generate_vector_deserializer(
                 result
             }
         }
-        ParsedType::Vector(ty) => {
-            let deserializer_name = format!("{}_{}", arg_name, ty);
+        ParsedType::Vector(ty, _) => {
+            let deserializer_name = format!("{}_{}", arg_name, ty)
+                .replace("&", "_")
+                .replace("<", "_")
+                .replace(">", "_");
```
Comment (folex) [inline crates/wit/src/parsed_type/vector_utils.rs:184]: Maybe move that to a separate function? I've seen same code in two more places

**Review by folex — APPROVED** [review]

**Pull request merged**